### ISSUE TITLE
[Question] Apply custom blendMode to PIXI particle effect

### ISSUE THREAD

**Issue by luludada** [pixijs/pixi.js#6798]
### Question
​
Hi, I am new to Pixi.js and I have been searching and looking for the custom blend mode usage with only WebGL in Pixi.js. I looked through the post(https://www.html5gamedevs.com/topic/30771-globalcompositeoperation-destination-in/) and the github PR (https://github.com/pixijs/pixi.js/pull/4049) to explore the possible to create my custom BlendMode. 

Here are my questions:

1. Does pixi v5 support the custom BlendMode like #4049 mentions, and it seems the version that above post uses is v4.5.3

2. I have no idea how to apply the custom BlendMode to the particle effect with reading the repo (https://github.com/pixijs/pixi-particles). I think I should make some change in ParticleUtils.ts, but I have no clue to begin.

3. I also looked the PR about more BlendMode(https://github.com/pixijs/pixi.js/pull/5391), and does this mean if I want to achieve the custom BlendMode,  it can only happen in canvas2D but not in WebGL?

4. If I can not use the method:

```javascript
var gl = WebGLRenderingContext;

var blend = new PIXI.BlendMode(gl.ZERO, gl.DST_ALPHA, gl.ONE, gl.ONE_MINUS_DST_ALPHA);

sprite.blendMode = blend;
```

 then how do I construct the shader of blendMode like ` blendFunc(gl.DST_ALPHA, gl.ONE_MINUS_DST_ALPHA)` ? 

Any advices are appreciated! Thanks in advanced!

- **`pixi.js` version**: v5.3.2
- **Browser & Version**: Chrome 84
- **OS & Version**: macOS 10.15

**Comment by andrewstart** [comment]
I can't speak to the creation of blend modes, but for particle effects, if the particle config says to use a blend mode `"my custom blend"` (or `"MY_CUSTOM_BLEND"` the library will use `PIXI.BlendModes.MY_CUSTOM_BLEND` (all uppercase, replace spaces with underscores). Alternatively, after initializing your emitter you can set `myEmitter.particleBlendMode = <custom blend mode value>` which will then be applied directly to each particle.

**Comment by ivanpopelyshev** [comment]
That PR wasnt merged. You can hack it the same way as in v4, its somewhere in "renderer.state" (StateSystem)

**Comment by luludada** [comment]
@ivanpopelyshev 

Hi Ivan, I am a bit confused about what you mean. 

To my understanding, should I find the code about renderer.state in that PR, and combine them into v5?

Or I should to achieve the custom blendMode in v5, should I follow the the repo pixi-picture (https://github.com/pixijs/pixi-picture) to create the shader?

**Comment by luludada** [comment]
@andrewstart
Hi Andrew, I am wondering if you have the tool to convert the plist file to pixi particle config since I need to compare the output of pixi custom blendMode and my particle plist file. Thanks in advanced.

**Comment by andrewstart** [comment]
I do not.

**Comment by ivanpopelyshev** [comment]
https://github.com/pixijs/pixi.js/blob/dev/packages/core/src/state/utils/mapWebGLBlendModesToPixi.ts
https://github.com/pixijs/pixi.js/blob/dev/packages/core/src/state/StateSystem.ts#L110

### Porter-Duff, like `gl.DST_ALPHA, gl.ONE_MINUS_DST_ALPHA` :

that means `renderer.state.blendModes`  holds all constants. You can add yours there under number that wasnt used before. Then assign `sprite.blendMode` to that number. **You dont have to modify pixiJS itself.** You can do it from outside.

### Custom blending mode:

Look https://github.com/pixijs/pixi-picture/blob/master/src/ShaderParts.ts#L128 , You can add your own code in `PIXI.picture.blends.blendFullArray[myBlendMode]`. 

### Summary

Learn how to hack stuff. You are missing the most powerful side of JavaScript.

**Comment by luludada** [comment]
@ivanpopelyshev 
Hi Ivan, I started including the pixi-picture.js in the html, and write the shader for SRC_ATOP BlendMode (array[BLEND_MODES.SRC_ATOP] = [gl.DST_ALPHA, gl.ONE_MINUS_SRC_ALPHA];). Refer to the "https://www.w3.org/TR/compositing-1/#porterduffcompositingoperators_srcatop", I am wondering about the Cs value of SRC_ATOP blend mode since SRC_ATOP is a compositing operator. My shader was having some errors, but I am not sure where did I do wrong on this. Thank you. 

My shader is:
```javascript
export const SRC_ATOP_FULL = 
        `{
            if (b_src.a == 0.0) {
                gl_FragColor = vec4(0, 0, 0, 0);
                return;
            }

            vec3 Cb = b_src.rgb / b_src.a;
            vec3 Cs;
            if (b_dest.a > 0.0) {
                Cs = b_dest.rgb / b_dest.a;
            }
            b_res.rgb = b_src.a * Cs * b_dest.a + b_dest.a * Cb * (1.0 - b_src.a);
            b_res.a = b_src.a * b_dest.a + b_dest.a * (1.0 - b_src.a);
         }
         `;

```

**Comment by ivanpopelyshev** [comment]
>  and write the shader for SRC_ATOP BlendMode

...

it already exists in pixi-v5.  https://github.com/pixijs/pixi.js/blob/dev/packages/core/src/state/utils/mapWebGLBlendModesToPixi.ts#L44

alpha is default, ONE, ONE_MINUS_SRC_ALPHA.

you are wrong to use 0 instead of 0.0 , integers are not floating points!

**Comment by ivanpopelyshev** [comment]
Just in case: if `PIXI.BLEND_MODES.SRC_ATOP` doesnt work for your case, but pixi-picture filter works, it means you want to use it on container that has multiple things, probably that will also work:

```
container.filters = [new PIXI.AlphaFilter()];
container.filters[0].blendMode = PIXI.BLEND_MODES.SRC_ATOP;
```

**Comment by luludada** [comment]
@ivanpopelyshev 
In my project, the particles will render at the same layer, but each particle will have different Z-Index. `PIXI.BLEND_MODES.SRC_ATOP` does not work for my case, the display of the particle effect is the same as the normal, so I decided to implement shader of SRC_ATOP. But my implement for shader of SRC_ATOP does not work in my project. I am not sure what you mean about I use 0 instead of 0.0 because I have not used 0 in anywhere. 

My shader now looks like:
```
{
            if (b_src.a == 0.0) {
                gl_FragColor = vec4(0, 0, 0, 0);
                return;
            }

            vec3 Cb = b_src.rgb / b_src.a, Cs;
            if (b_dest.a > 0.0) {
                Cs = b_dest.rgb / b_dest.a;
            }

            if (b_dest.a > 0.0) {
                b_res.rgb = b_src.a * Cs * b_dest.a + b_dest.a * Cb * (1.0 - b_src.a);
                b_res.a = b_src.a * b_dest.a + b_dest.a * (1.0 - b_src.a);
            }
         }
```

But I kept getting two main errors: one is `FilterState` miss the constructor, and another one is `framebuffer` is undefined and cannot access the `state.renderTexture.framebuffer.multisample`. 
For the FilterState constructor, I checked your old commit history, I found you deleted it. I am not sure without the FilterState class declaration in `FilterSystemMixin.ts`, where I can find the class `FilterState` declaration in PIXI. For the framebuffer error, I have not figured out what happened.

**Comment by luludada** [comment]
@ivanpopelyshev 

I am using pixi.js v5.1.5, and I just updated to v5.3.2 now. 

I am not sure if you heared about cocos creator. I am trying to render the particle effect in pixi like this:
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/29498030/91163646-82481600-e700-11ea-9444-62d0ebc21cc6.gif)

But if use 

```javascript
container.filters = [new PIXI.AlphaFilter()];
container.filters[0].blendMode = PIXI.BLEND_MODES.SRC_ATOP;
```
The output looks either like this or the  particle effect does not render at all, and  SRC_ATOP seems not work out.  

![ezgif com-video-to-gif (1)](https://user-images.githubusercontent.com/29498030/91164193-43669000-e701-11ea-968f-56c481adf19b.gif)

I compared this effect with cocos creator, I found this effect in cocos2d is called [gl.SRC_ALPHA, gl.ONE_MINUS_SRC_ALPHA], then I knew that coocs creator is using not premultiply alpha.
